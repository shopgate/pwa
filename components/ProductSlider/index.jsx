import React from 'react';
import PropTypes from 'prop-types';
import { Swiper } from '@shopgate/engage/components';
import { Theme } from '@shopgate/pwa-common/context';
import styles from './style';

/**
 * @param {Object} props The component props.
 * @returns {JSX}
 */
function ProductSlider(props) {
  const {
    autoplay,
    className,
    delay,
    productIds,
    slidesPerView,
    snap,
  } = props;

  return (
    <Theme>
      {({ ProductCard }) => {
        const Item = props.item || ProductCard;
        return (
          <Swiper
            autoPlay={autoplay}
            className={className}
            controls={false}
            indicators={false}
            interval={delay}
            loop={false}
            freeMode={!snap}
            slidesPerView={slidesPerView}
          >
            {productIds.map(id => (
              <Swiper.Item key={id}>
                <Item productId={id} style={styles} />
              </Swiper.Item>
            ))}
          </Swiper>
        );
      }}
    </Theme>
  );
}

ProductSlider.propTypes = {
  productIds: PropTypes.arrayOf(PropTypes.string).isRequired,
  autoplay: PropTypes.bool,
  className: PropTypes.string,
  delay: PropTypes.number,
  item: PropTypes.func,
  slidesPerView: PropTypes.number,
  snap: PropTypes.bool,
};

ProductSlider.defaultProps = {
  autoplay: false,
  className: null,
  delay: 10,
  item: null,
  slidesPerView: 2.3,
  snap: false,
};

export default ProductSlider;
