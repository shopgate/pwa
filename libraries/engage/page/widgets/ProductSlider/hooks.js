import { useMemo } from 'react';
import { camelCase } from 'lodash';
import { useWidget } from '@shopgate/engage/page/hooks';

/**
 * @typedef {Object} ProductSliderWidgetProducts
 * @property {"searchTerm" | "brand" | "category" | "itemNumbers"} productSelectorType Source type
 * for the product list.
 * @property {string} productsSearchTerm A search term to filter products by
 * @property {string} productsBrand A brand to filter products by
 * @property {string} productsCategory A category to filter products by
 * @property {string} productsItemNumbers A comma-separated list of item numbers to filter products
 * by
 */

/**
 * @typedef {Object} ProductSliderWidgetConfig
 * @property {ProductSliderWidgetProducts} products The products configuration for the widget.
 * @property {number} productCount The number of products to display in the widget
 * @property {"relevance" | "PriceDesc" | "PriceDesc"} sort Sort order for the products
 * @property {boolean} loadMoreButton Whether to display a "Load more" button
 * @property {boolean} showName Whether to display product names
 * @property {boolean} showPrice Whether to display product prices
 * @property {boolean} showRating Whether to display product ratings
 */

/**
 * @typedef {ReturnType< typeof import('@shopgate/engage/page/hooks')
 * .useWidget<ProductSliderWidgetConfig> >} UseWidgetReturnType
 */

/**
 * Hook to access the Product Slider widget configuration.
 * @returns {Object} the widget config
 */
export const useProductSliderWidget = () => {
  /** @type {UseWidgetReturnType}  */
  const { config } = useWidget();

  const {
    products,
    productCount,
    sort,
    showName = false,
    showPrice = false,
    showRating = false,
    slideAutomatic = true,
    endlessSlider = true,
    sliderSpeed = 7000,
  } = config;

  const {
    productSelectorType,
    productsBrand,
    productsCategory,
    productsItemNumbers,
    productsSearchTerm,
  } = products;

  const value = useMemo(() => {
    switch (productSelectorType) {
      case 'brand':
        return productsBrand;
      case 'category':
        return productsCategory;
      case 'itemNumbers':
        return productsItemNumbers.split(',').map(item => item.trim());
      case 'searchTerm':
      default:
        return productsSearchTerm;
    }
  }, [
    productSelectorType,
    productsBrand,
    productsCategory,
    productsItemNumbers,
    productsSearchTerm,
  ]);

  const flags = useMemo(() => ({
    hideName: !showName,
    hidePrice: !showPrice,
    hideRating: !showRating,
    autoplay: slideAutomatic,
    delay: sliderSpeed,
    loop: endlessSlider,
  }), [showName, showPrice, showRating, slideAutomatic, sliderSpeed, endlessSlider]);

  return {
    productsSearchType: productSelectorType,
    productsSearchValue: value,
    sort: camelCase(sort),
    productCount,
    flags,
  };
};
