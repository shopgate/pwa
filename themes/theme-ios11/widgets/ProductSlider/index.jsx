import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Slider from '@shopgate/pwa-common/components/Slider';
import Card from '@shopgate/pwa-ui-shared/Card';
import ProductCard from 'Components/ProductCard';
import { transformDisplayOptions } from '@shopgate/pwa-common/helpers/data';
import connect from './connector';
import styles from './style';

/**
 * Creates an item for a single product.
 * @param {Object} product The product data.
 * @param {boolean} settings.showName Show the product name?
 * @param {boolean} settings.showPrice Show the product price?
 * @param {boolean} settings.showReviews Show the product reviews?
 * @return {JSX} The rendered product card.
 */
const createSliderItem = (product, { showName, showPrice, showReviews }) => {
  const key = `s${product.id}`;

  return (
    <Slider.Item key={key} className={styles.sliderItem}>
      <Card className={styles.card}>
        <ProductCard
          product={product}
          hideName={!showName}
          hidePrice={!showPrice}
          hideRating={!showReviews}
          titleRows={2}
        />
      </Card>
    </Slider.Item>
  );
};

/**
 * The core product slider widget.
 */
class ProductSlider extends Component {
  static propTypes = {
    getProducts: PropTypes.func.isRequired,
    id: PropTypes.string.isRequired,
    // The settings as received by the pipeline request
    settings: PropTypes.shape({
      headline: PropTypes.string.isRequired, // The headline of the product slider.
      layout: PropTypes.string.isRequired, // The layout of the product slider.
      queryParams: PropTypes.oneOfType([ // The query parameters for this slider.
        PropTypes.string,
        PropTypes.array,
      ]),
      queryType: PropTypes.number.isRequired, // The query type ID.
      showName: PropTypes.bool.isRequired, // Should the product name be displayed?
      showPrice: PropTypes.bool.isRequired, // Should the product price be displayed?
      showReviews: PropTypes.bool.isRequired, // Should the user ratings be displayed?
      sortOrder: PropTypes.string.isRequired, // The sort order.
      sliderSettings: PropTypes.shape({
        autostart: PropTypes.bool.isRequired, // Should the slider start automatically?
        delay: PropTypes.oneOfType([ // The delay between the automatic slides loops
          PropTypes.string,
          PropTypes.number,
        ]).isRequired,
      }).isRequired,
    }).isRequired,
    products: PropTypes.arrayOf(PropTypes.shape()),
  };

  static defaultProps = {
    products: [],
  };

  /**
   * Called when the component is mounted, requests the products.
   */
  componentDidMount() {
    const { getProducts, id } = this.props;
    const { queryType, queryParams } = this.props.settings;

    getProducts(
      queryType,
      queryParams,
      {
        sort: transformDisplayOptions(this.props.settings.sortOrder),
      },
      id
    );
  }

  /**
   * Renders a headline if we have one.
   * @returns {JSX}
   */
  renderHeadline = () => {
    if (this.props.settings.headline) {
      return (
        <h3 className={styles.headline}>
          {this.props.settings.headline}
        </h3>
      );
    }

    return null;
  };

  /**
   * Renders the widget.
   * @return {JSX}
   */
  render() {
    const { settings } = this.props;
    const { sliderSettings } = settings;

    // Create the slides for each product, only displays the first 30 products.
    const items = this.props.products.slice(0, 30).map((
      product => createSliderItem(product, settings)
    ));

    if (!items.length) {
      return null;
    }

    // Finally, build the slider.
    return (
      <div className={styles.slider}>
        {this.renderHeadline()}
        <Slider
          autoPlay={sliderSettings.autostart}
          loop={false}
          indicators={false}
          controls={false}
          interval={Number.parseInt(sliderSettings.delay, 10)}
          snapItems={false}
          slidesPerView={2.3}
          classNames={{ container: styles.sliderContainer }}
        >
          {items}
        </Slider>
      </div>
    );
  }
}

export default connect(ProductSlider);
