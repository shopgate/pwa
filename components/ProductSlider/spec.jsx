/* eslint-disable require-jsdoc */
import React from 'react';
import { shallow } from 'enzyme';
import ProductSlider from './index';

jest.mock('@shopgate/engage/components', () => {
  const Swiper = ({ children }) => children;
  Swiper.Item = ({ children }) => children;
  Swiper.Item.displayName = 'Swiper.Item';
  return {
    Swiper,
  };
});
jest.mock('@shopgate/pwa-common/context', () => {
  const ProductCard = props => <ProductCard {...props} />;
  return {
    Theme: ({ children }) => children({ ProductCard }),
  };
});

describe('<ProductSlider />', () => {
  it('should match snapshot', () => {
    const wrapper = shallow((
      <ProductSlider productIds={['prod1']} />
    )).dive();

    expect(wrapper).toMatchSnapshot();
  });
});
/* eslint-enable require-jsdoc */
