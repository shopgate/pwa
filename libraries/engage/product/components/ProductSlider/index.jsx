import React from 'react';
import PropTypes from 'prop-types';
import { useWidgetSettings } from '@shopgate/engage/core';
import { Swiper } from '@shopgate/engage/components';
import { Theme } from '@shopgate/engage/core/contexts';
import {
  ProductListTypeProvider,
  ProductListEntryProvider,
} from '@shopgate/engage/product/providers';
import { container, items } from './style';

export const WIDGET_ID = '@shopgate/engage/product/ProductSlider';

/**
 * @param {Object} props The component props.
 * @returns {JSX}
 */
function ProductSlider(props) {
  const {
    autoplay,
    className,
    delay,
    productIds,
    snap,
    scope,
  } = props;

  const widgetSettings = useWidgetSettings(WIDGET_ID) || {};
  const { slidesPerView = 2.3 } = props.slidesPerView ? props : widgetSettings;

  return (
    <Theme>
      {({ ProductCard }) => {
        const Item = props.item || ProductCard;
        return (
          <ProductListTypeProvider type="productSlider" subType={scope}>
            <Swiper
              autoPlay={autoplay}
              className={`${className} engage__product__product-slider`}
              controls={false}
              indicators={false}
              interval={delay}
              loop={false}
              freeMode={!snap}
              slidesPerView={slidesPerView}
            >
              {productIds.map(id => (
                <Swiper.Item key={id} className={container}>
                  <ProductListEntryProvider productId={id}>
                    <Item productId={id} style={items} />
                  </ProductListEntryProvider>
                </Swiper.Item>
              ))}
            </Swiper>
          </ProductListTypeProvider>
        );
      }}
    </Theme>
  );
}

ProductSlider.WIDGET_ID = WIDGET_ID;

ProductSlider.propTypes = {
  productIds: PropTypes.arrayOf(PropTypes.string).isRequired,
  autoplay: PropTypes.bool,
  className: PropTypes.string,
  delay: PropTypes.number,
  item: PropTypes.func,
  /**
   * Optional scope of the component. Will be used as subType property of the ProductListTypeContext
   * and is intended as a description in which "context" the component is used.
   * @default null
   */
  scope: PropTypes.string,
  slidesPerView: PropTypes.number,
  snap: PropTypes.bool,
};

ProductSlider.defaultProps = {
  autoplay: false,
  className: null,
  delay: 10,
  item: null,
  slidesPerView: null,
  snap: false,
  scope: null,
};

export default ProductSlider;
