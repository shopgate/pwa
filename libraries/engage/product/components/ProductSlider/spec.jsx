/* eslint-disable require-jsdoc */
import React from 'react';
import { shallow } from 'enzyme';
import ProductSlider from './index';

jest.mock('@shopgate/engage/core', () => ({
  useWidgetSettings: jest.fn(),
}));

jest.mock('@shopgate/engage/components', () => {
  const Swiper = ({ children }) => children;
  Swiper.Item = ({ children }) => children;
  Swiper.Item.displayName = 'Swiper.Item';
  return {
    Swiper,
  };
});

jest.mock('@shopgate/engage/core/hooks', () => {
  const ProductCard = props => <ProductCard {...props} />;
  return {
    useThemeComponents: () => ({ ProductCard }),
  };
});

describe('<ProductSlider />', () => {
  it('should match snapshot', () => {
    const wrapper = shallow((
      <ProductSlider productIds={['prod1']} />
    ));

    expect(wrapper).toMatchSnapshot();
  });
});
/* eslint-enable require-jsdoc */
