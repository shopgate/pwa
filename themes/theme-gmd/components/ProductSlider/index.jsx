import React from 'react';
import PropTypes from 'prop-types';
import { Slider } from '@shopgate/pwa-common/components';
import { Theme } from '@shopgate/pwa-common/context';
import styles from './style';

/**
 * @param {Object} props The component props.
 * @returns {JSX}
 */
function ProductSlider(props) {
  const {
    autoplay,
    className,
    delay,
    productIds,
    slidesPerView,
    snap,
  } = props;

  return (
    <Theme>
      {({ ProductCard }) => {
        const Item = props.item || ProductCard;
        return (
          <Slider
            autoPlay={autoplay}
            className={className}
            controls={false}
            indicators={false}
            interval={delay}
            loop={false}
            snapItems={snap}
            slidesPerView={slidesPerView}
          >
            {productIds.map(id => <Item key={id} productId={id} style={styles} />)}
          </Slider>
        );
      }}
    </Theme>
  );
}

ProductSlider.propTypes = {
  productIds: PropTypes.arrayOf(PropTypes.string).isRequired,
  autoplay: PropTypes.bool,
  className: PropTypes.string,
  delay: PropTypes.number,
  item: PropTypes.func,
  slidesPerView: PropTypes.number,
  snap: PropTypes.bool,
};

ProductSlider.defaultProps = {
  autoplay: false,
  className: null,
  delay: 10,
  item: null,
  slidesPerView: 2.3,
  snap: false,
};

export default ProductSlider;
