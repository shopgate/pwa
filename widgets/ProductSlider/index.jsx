import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Swiper, Card } from '@shopgate/engage/components';
import { ProductCard } from '@shopgate/engage/product';
import { transformDisplayOptions } from '@shopgate/pwa-common/helpers/data';
import { withWidgetSettings } from '@shopgate/engage/core';
import { WIDGET_ID } from 'Components/ProductSlider';
import Headline from 'Components/Headline';
import connect from './connector';
import styles from './style';

/**
 * The core product slider widget.
 */
class ProductSlider extends PureComponent {
  static propTypes = {
    getProducts: PropTypes.func.isRequired,
    id: PropTypes.string.isRequired,
    // The settings as received by the pipeline request
    settings: PropTypes.shape({
      headline: PropTypes.string.isRequired, // The headline of the product slider.
      layout: PropTypes.string.isRequired, // The layout of the product slider.
      queryParams: PropTypes.oneOfType([ // The query parameters for this slider.
        PropTypes.string,
        PropTypes.array,
      ]),
      queryType: PropTypes.number.isRequired, // The query type ID.
      showName: PropTypes.bool.isRequired, // Should the product name be displayed?
      showPrice: PropTypes.bool.isRequired, // Should the product price be displayed?
      showReviews: PropTypes.bool.isRequired, // Should the user ratings be displayed?
      sortOrder: PropTypes.string.isRequired, // The sort order.
      sliderSettings: PropTypes.shape({
        autostart: PropTypes.bool.isRequired, // Should the slider start automatically?
        delay: PropTypes.oneOfType([ // The delay between the automatic slides loops
          PropTypes.string,
          PropTypes.number,
        ]).isRequired,
      }).isRequired,
    }).isRequired,
    products: PropTypes.arrayOf(PropTypes.shape()),
    widgetSettings: PropTypes.shape({
      slidesPerView: PropTypes.number,
    }),
  };

  static defaultProps = {
    products: [],
    widgetSettings: {},
  };

  /**
   * Called when the component is mounted, requests the products.
   */
  componentDidMount() {
    const { getProducts, id } = this.props;
    const { queryType, queryParams } = this.props.settings;

    getProducts(
      queryType,
      queryParams,
      {
        sort: transformDisplayOptions(this.props.settings.sortOrder),
      },
      id
    );
  }

  /**
   * Renders a headline if we have one.
   * @returns {JSX}
   */
  renderHeadline = () => {
    if (this.props.settings.headline) {
      return (
        <Headline text={this.props.settings.headline} />
      );
    }

    return null;
  };

  /**
   * Renders the widget.
   * @return {JSX}
   */
  render() {
    const { settings, products, widgetSettings } = this.props;
    const {
      sliderSettings, showName, showPrice, showReviews,
    } = settings;
    const { slidesPerView = 2.3 } = widgetSettings;

    if (!products.length) {
      return null;
    }

    // Finally, build the slider.
    return (
      <div className={styles.slider}>
        {this.renderHeadline()}
        <Swiper
          autoPlay={sliderSettings.autostart}
          loop={false}
          indicators={false}
          controls={false}
          interval={Number.parseInt(sliderSettings.delay, 10)}
          freeMode
          slidesPerView={slidesPerView}
          classNames={{ container: styles.sliderContainer }}
        >
          {products.slice(0, 30).map(product => (
            <Swiper.Item key={product.id} className={styles.sliderItem}>
              <Card className={styles.card}>
                <ProductCard
                  product={product}
                  hideName={!showName}
                  hidePrice={!showPrice}
                  hideRating={!showReviews}
                  titleRows={2}
                />
              </Card>
            </Swiper.Item>
          ))}
        </Swiper>
      </div>
    );
  }
}

export default withWidgetSettings(connect(ProductSlider), WIDGET_ID);

export { ProductSlider as UnwrappedProductSlider };
