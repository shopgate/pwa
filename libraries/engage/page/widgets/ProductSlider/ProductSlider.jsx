import React, { useMemo } from 'react';
import { ProductSlider } from '@shopgate/engage/product/components';
import { useWidgetProducts } from '@shopgate/engage/page/hooks';
import { useTheme } from '@shopgate/engage/styles';
import { useProductSliderWidget } from './hooks';
import WidgetHeadline from '../../components/WidgetHeadline';

/**
 * The ProductSliderWidget is used to display a product slider.
 * @returns {JSX.Element}
 */
const ProductSliderWidget = () => {
  const {
    productsSearchType,
    productsSearchValue,
    sort,
    productCount,
    swiperProps,
    productItemProps,
    isPreview,
    showHeadline,
    headline,
  } = useProductSliderWidget();

  const {
    results,
  } = useWidgetProducts({
    type: productsSearchType,
    value: productsSearchValue,
    limit: productCount,
    sort,
  });
  const theme = useTheme();
  const productIds = useMemo(() => results?.map(result => result.id), [results]);

  if (!productIds || !productIds.length) {
    return null;
  }

  return (
    <>
      {(showHeadline && headline && productIds.length) ? (
        <WidgetHeadline headline={headline} />
      ) : null}
      <ProductSlider
        productIds={productIds}
        scope="widgets"
        productItemProps={productItemProps}
        slidesPerView={2.3}
      // Improves interaction with the slider in the CMS preview iframe
        {...isPreview ? { touchStartPreventDefault: true } : {}}
        breakpoints={{
          [theme.breakpoints.values.sm]: {
            slidesPerView: 3.3,
          },
          [theme.breakpoints.values.md]: {
            slidesPerView: 4.3,
          },
          [theme.breakpoints.values.lg]: {
            slidesPerView: 5.3,
          },
        }}
        {...swiperProps}
      />
    </>
  );
};

export default ProductSliderWidget;
