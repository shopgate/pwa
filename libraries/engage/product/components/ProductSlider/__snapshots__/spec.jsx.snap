// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductSlider /> should match snapshot 1`] = `
<ProductListTypeProvider
  subType={null}
  type="productSlider"
>
  <Swiper
    autoPlay={false}
    className="null engage__product__product-slider"
    controls={false}
    freeMode={true}
    indicators={false}
    interval={10}
    loop={false}
    slidesPerView={2.3}
  >
    <Swiper.Item
      className="css-90ryaa"
      key="prod1"
    >
      <ProductCard
        productId="prod1"
        style={
          Object {
            "height": "100%",
            "margin": "0 8px",
          }
        }
      />
    </Swiper.Item>
  </Swiper>
</ProductListTypeProvider>
`;
